### PR TITLE
Search bar for GPU Commands

### PR DESCRIPTION
**Pull request by RiokuTheSlayer**
Search bar for the Pica Command List. Just adds a text area that updates the GPU command list with the input from the user. Case-sensitive. Needs optimized, pretty slow, but there's a lot of commands so that may be unavoidable. Spacing should be fixed now, I un-tabbed all my changes.
(As a side note I moved the search bar up above the start/finish/copy trace buttons. Screenshot isn't updated.)

Screenshot of the change in Citra:
screenshot 28

### CODE REVIEW

**Comment by neobrain** [comment]
Indentation is completely messed up with these changes, both violating our coding style and furthermore making the diff view useless. Please fix this up so that we can do a proper review!

**Comment by neobrain** [comment]
Your most recent patch notes say "Spacing/indentation should be 100% valid now". In reply to that statement, would you mind checking the `git diff --color=auto` output before pushing changes?

File: src/video_core/debug_utils/debug_utils.h [hunk 1]
```diff
@@ -47,15 +47,15 @@ class DebugContext {
...
-     */
+    * Inherit from this class to be notified of events registered to some debug context.
+    * Most importantly this is used for our debugger GUI.
+    *
+    * To implement event handling, override the OnPicaBreakPointHit and OnPicaResume methods.
+    * @warning All BreakPointObservers need to be on the same thread to guarantee thread-safe state
+    * access
+    * @todo Evaluate an alternative interface, in which there is only one managing observer and
+    * multiple child observers running (by design) on the same thread.
+    */
```
Comment (MerryMage) [inline src/video_core/debug_utils/debug_utils.h]: Unnecessary whitespace diff.

File: src/video_core/debug_utils/debug_utils.h [hunk 2]
```diff
@@ -80,50 +80,50 @@ class DebugContext {
...
-         * Action to perform when a breakpoint was reached.
-         * @param event Type of event which triggered the breakpoint
-         * @param data Optional data pointer (if unused, this is a nullptr)
-         * @note This function will perform nothing unless it is overridden in the child class.
-         */
+        * Action to perform when a breakpoint was reached.
+        * @param event Type of event which triggered the breakpoint
+        * @param data Optional data pointer (if unused, this is a nullptr)
+        * @note This function will perform nothing unless it is overridden in the child class.
+        */
```
Comment (MerryMage) [inline src/video_core/debug_utils/debug_utils.h]: Unnecessary whitespace diff.

File: src/video_core/debug_utils/debug_utils.h [hunk 3]
```diff
@@ -80,50 +80,50 @@ class DebugContext {
...
         virtual void OnPicaBreakPointHit(Event, void*) {}
 
         /**
-         * Action to perform when emulation is resumed from a breakpoint.
-         * @note This function will perform nothing unless it is overridden in the child class.
-         */
+        * Action to perform when emulation is resumed from a breakpoint.
+        * @note This function will perform nothing unless it is overridden in the child class.
+        */
```
Comment (MerryMage) [inline src/video_core/debug_utils/debug_utils.h]: Unnecessary whitespace diff; ditto for all doc comments in this file.

File: src/video_core/debug_utils/debug_utils.h [hunk 4]
```diff
@@ -267,4 +298,4 @@ class MemoryAccessTracker {
 
 } // namespace
 
-} // namespace
+} // namespace
```
Comment (MerryMage) [inline src/video_core/debug_utils/debug_utils.h]: New line required at end of file.

**Comment by RiokuTheSlayer** [comment]
That's so weird because I'm using clang-format to do everything, the latest Visual Studio version. So... I'll go remove them and stop using the plugin I guess

File: src/video_core/debug_utils/debug_utils.h [hunk 5]
```diff
@@ -194,6 +194,37 @@ struct PicaTrace {
...
+            is_search_corrected = true;
+
+        for (int i = 0; i < writes.size(); i++) {
+            bool is_valid = true;
+            std::string cmd_name = Pica::Regs::GetCommandName(writes[i].cmd_id);
+
+            if (cmd_name.size() < search_string.size())
+                continue;
+
+            for (int j = 0; j < search_string.size(); j++) {
```
Comment (MerryMage) [inline src/video_core/debug_utils/debug_utils.h]: Consider `std::equal(cmd_name.begin(), cmd_name.end(), search_string.begin());` instead of this `for` loop.

File: src/video_core/debug_utils/debug_utils.h [hunk 6]
```diff
@@ -194,6 +194,37 @@ struct PicaTrace {
...
+                    is_valid = false;
+                    break;
+                }
+            } // End of j for-loop
+
+            if (!is_valid)
+                continue;
+
+            search_corrected_writes.push_back(writes[i]);
+        } // End of i for-loop
```
Comment (MerryMage) [inline src/video_core/debug_utils/debug_utils.h]: Consider

    auto matches_search_string = [&](const Write& write) -> bool {
        std::string cmd_name = Pica::Regs::GetCommandName(write.cmd_id);
        return cmd_name.size() < search_string.size() &&
            std::equal(cmd_name.begin(), cmd_name.end(), search_string.begin());
    }
    std::copy_if(writes.begin(), write.end(), std::back_inserter(search_corrected_writes), matches_search_string);

or equivalent instead of this `for` loop.

File: src/video_core/debug_utils/debug_utils.h (hunk 5 above)
Comment (neobrain) [inline src/video_core/debug_utils/debug_utils.h]: Even with @MerryMage's suggestion, this only looks for prefix matches of the entered string. For a search function, you'd expect any infixes to match.

Alternative suggestion for replacing the entire `for (int i = 0; i < writes.size(); ++i)` loop:
```
auto&& is_matching_write = [&search_string](auto&& write) {
    auto cmd_name = Pica::Regs::GetCommandName(write.cmd_id);
    return cmd_name.end() != boost::search(cmd_name, search_string);
};

boost::copy(writes | boost::adaptors::filtered(is_matching_write),
            std::back_inserter(search_corrected_writes));
```

File: src/video_core/debug_utils/debug_utils.h [hunk 7]
```diff
@@ -194,6 +194,37 @@ struct PicaTrace {
         u32 value;
     };
     std::vector<Write> writes;
+    std::vector<Write> search_corrected_writes;
+    bool is_search_corrected = false;
+
+    void CorrectSearchWrites(std::string search_string) {
+        search_corrected_writes.clear();
+
+        if (search_string.size() == 0)
```
Comment (neobrain) [inline src/video_core/debug_utils/debug_utils.h]: `is_search_corrected = (search_string.size() > 0);`

**Comment by wwylele** [comment]
What's the status of this PR?

**Comment by yuriks** [comment]
I like the idea, but instead of outright filtering (which removes the context from the things you were searching for) how about making it highlight entries which match the string? That way you can easily find them in the list, but you can still see what commands are used together. Then pressing Enter in the textbox could can bring you to the next result.

**Comment by Subv** [comment]
I think this has been sitting around for enough time, if you still plan on continuing it, please open a new PR with the requested changes.
